### PR TITLE
[framework] remove casts from RO parent traversal

### PR DESCRIPTION
**Pull request by jonahwilliams**
By making AbstractTree generic, we can remove _most_ of the casts - especially those of the form:

```
while (condition) 
  renderObject = this.parent as RenderObject;
```

There is still one cast required when adopting a child - but this is really just moved from the covariant cast on the method body. This breaks a test which claims that RenderObjects can have non RenderObject parents.

https://github.com/flutter/flutter/issues/97310

### CODE REVIEW

File: packages/flutter/lib/src/foundation/node.dart [hunk 1]
```diff
@@ -115,25 +116,25 @@ class AbstractNode {
...
     assert(() {
-      AbstractNode node = this;
+      AbstractNode<T> node = this;
       while (node.parent != null)
         node = node.parent!;
       assert(node != child); // indicates we are about to create a cycle
       return true;
     }());
-    child._parent = this;
+    child._parent = this as T;
```
Comment (jonahwilliams) [inline packages/flutter/lib/src/foundation/node.dart:137]: This is really just moving the cast from L127 to here

**Comment by jonahwilliams** [comment]
Specifically this causes:

```
03:30 +5839 ~14 -1: /b/s/w/ir/x/w/flutter/packages/flutter/test/rendering/non_render_object_root_test.dart: non-RenderObject roots [E]                                                                 
  type 'RealRoot' is not a subtype of type 'RenderObject' in type cast
  package:flutter/src/foundation/node.dart 137:26        AbstractNode.adoptChild
  test/rendering/non_render_object_root_test.dart 13:5   new RealRoot
  test/rendering/non_render_object_root_test.dart 46:27  main.<fn>
  ```
  
  To fail, since as typed it cannot adopt a child of a different type.

File: packages/flutter/lib/src/foundation/node.dart [hunk 2]
```diff
@@ -115,25 +116,25 @@ class AbstractNode {
   }
 
   /// The parent of this node in the tree.
-  AbstractNode? get parent => _parent;
-  AbstractNode? _parent;
+  T? get parent => _parent;
+  T? _parent;
```
Comment (Hixie) [inline packages/flutter/lib/src/foundation/node.dart:120]: I don't think that's right. There's no way for the class to know what it's _parent_ type will be (and it certainly doesn't need to be the same as the child type). RenderObject in particular is happy to have a parent that's not a RenderObject.

File: packages/flutter/lib/src/rendering/box.dart [hunk 3]
```diff
@@ -2149,7 +2149,7 @@ abstract class RenderBox extends RenderObject {
     assert(!_debugDoingBaseline, 'Please see the documentation for computeDistanceToActualBaseline for the required calling conventions of this method.');
     assert(!debugNeedsLayout);
     assert(() {
-      final RenderObject? parent = this.parent as RenderObject?;
+      final RenderObject? parent = this.parent;
```
Comment (Hixie) [inline packages/flutter/lib/src/rendering/box.dart:2152]: if we could make this change, we could also make this `this.parent!` and avoid the `!`s below, fwiw

Comment (Hixie) [reply packages/flutter/lib/src/rendering/box.dart:2152]: oh never mind i'm wrong.

File: packages/flutter/lib/src/foundation/node.dart (hunk 2 above)
Comment (Hixie) [reply packages/flutter/lib/src/foundation/node.dart:120]: If the performance improvement here is significant, we could make an explicit decision to change this design, and require that RenderObject trees not have a `parent` pointing to a non-RenderObject parent. If we do that we should update documentation and logic. Probably need to carefully read every mention of `parent` at least in `object.dart`, maybe other places like `binding.dart`, `view.dart`, and maybe `widgets/framework.dart`.

Comment (Hixie) [reply packages/flutter/lib/src/foundation/node.dart:120]: cc @goderbauer for thoughts on making such a radical change to AbstractNode's design.

Comment (jonahwilliams) [reply packages/flutter/lib/src/foundation/node.dart:120]: Yeah, that is the unfortunate part about this PR - to remove the cast from parent we'd need to enforce that each AbstractTree subtype knows its parent type.

Comment (jonahwilliams) [reply packages/flutter/lib/src/foundation/node.dart:120]: while I'm at it ... I wonder If I could also enforce that parent is non-null except for root.

Comment (goderbauer) [reply packages/flutter/lib/src/foundation/node.dart:120]: > thoughts on making such a radical change to AbstractNode's design.

We'd have to to an audit, but I don't have any immediately obvious concerns.

It does seem to clean up the code a little bit. I do wonder, though, how significant this change is for performance.

Comment (jonahwilliams) [reply packages/flutter/lib/src/foundation/node.dart:120]: We walk parent render objects in a few places, finding out the improvement would rely on finding a scenario where those functions are used heavily.

Comment (jonahwilliams) [reply packages/flutter/lib/src/foundation/node.dart:120]: i.e. getTransformTo, scroll offset reveal, et cetera

**Comment by jonahwilliams** [comment]
So I've been looking for a good way to benchmark this and found the existing suite too limiting for small performance improvements. I am working on an update to the test/benchmark code to support easier microbenchmarks

**Comment by jonahwilliams** [comment]
I continued investigating this for a while and  don't think its an improvement (or at least I didn't find a benchmark to show it). I'll close for now and follow up with some different experiments

**Pull request closed without merging**